compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -313,6 +313,20 @@ impl<'a> Parser<'a> {
                 break;
             }
 
+            // recovery from `if b && if let Some(x)`
+            if op == AssocOp::LAnd
+                && self.token.is_keyword(kw::If)
+                && self.look_ahead(1, |t| t.is_keyword(kw::Let))
+                && self
+                    .look_ahead(2, |t| t.is_ident_named(sym::Some) || t.is_ident_named(sym::None))
```

Comment:
The following code is perfectly valid:
```rust
if true
    && if let Some(1) = Some(1) { true } else { true }
{
    todo!()
}
```
In general, it is always possible to write valid code containing `if <some condition> && if let <some pattern>`.

I don't think it's possible to detect the scenario you're looking for until _after_ you parse the inner `if let ..` expression.